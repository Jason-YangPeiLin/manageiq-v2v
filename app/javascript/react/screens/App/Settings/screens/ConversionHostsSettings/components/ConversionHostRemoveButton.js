import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';

const ConversionHostRemoveButton = ({
  hostOrTask,
  setHostToDeleteAction,
  showConversionHostDeleteModalAction,
  ...props
}) => (
  <Button
    id={`remove_${hostOrTask.id}`}
    onClick={e => {
      e.stopPropagation();
      setHostToDeleteAction(hostOrTask);
      showConversionHostDeleteModalAction();
    }}
    {...props}
  >
    {__('Remove')}
  </Button>
);

ConversionHostRemoveButton.propTypes = {
  hostOrTask: PropTypes.object,
  setHostToDeleteAction: PropTypes.func,
  showConversionHostDeleteModalAction: PropTypes.func
};

export default ConversionHostRemoveButton;
