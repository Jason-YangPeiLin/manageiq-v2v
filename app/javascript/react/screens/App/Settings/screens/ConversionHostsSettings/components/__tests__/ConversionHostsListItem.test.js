import React from 'react';
import { shallow } from 'enzyme';
import { DropdownKebab, MenuItem } from 'patternfly-react';
import ConversionHostsListItem from '../ConversionHostsListItem';
import { success, inProgress, failed } from '../../../../settings.fixtures';
import ConversionHostRemoveButton from '../ConversionHostRemoveButton';
import ConversionHostRetryButton from '../ConversionHostRetryButton';

const conversionHostEnabled = {
  name: 'Conversion Host 1',
  id: '1',
  vddk_transport_supported: true,
  meta: {
    tasksByOperation: {
      enable: [
        {
          id: '1',
          meta: { isTask: true, operation: 'enable' },
          updated_on: '2019-05-13T18:06:11.849Z',
          ...failed
        },
        {
          id: '2',
          meta: { isTask: true, operation: 'enable', resourceName: 'Conversion Host 1' },
          updated_on: '2019-05-13T18:07:11.849Z',
          context_data: {
            conversion_host_enable: 'MOCK PLAYBOOK LOG OUTPUT',
            conversion_host_check: 'MOCK PLAYBOOK LOG OUTPUT'
          },
          ...success
        }
      ]
    }
  }
};
const conversionHostDisabling = {
  name: 'Conversion Host 1',
  id: '1',
  vddk_transport_supported: true,
  meta: {
    tasksByOperation: {
      enable: [
        {
          id: '1',
          meta: { isTask: true, operation: 'enable' },
          updated_on: '2019-05-13T18:05:11.849Z',
          ...success
        }
      ],
      disable: [
        {
          id: '3',
          meta: { isTask: true, operation: 'disable' },
          updated_on: '2019-05-13T18:07:11.849Z',
          ...inProgress
        },
        {
          id: '2',
          meta: { isTask: true, operation: 'disable' },
          updated_on: '2019-05-13T18:06:11.849Z',
          context_data: { request_params: { mock: 'params' } },
          ...failed
        }
      ]
    }
  }
};
const enableTaskInProgress = {
  id: '1',
  meta: { isTask: true, operation: 'enable' },
  context_data: { request_params: { vmware_vddk_package_url: 'foo' } },
  ...inProgress
};
const enableTaskFailed = {
  id: '1',
  meta: { isTask: true, operation: 'enable' },
  context_data: { request_params: { vmware_vddk_package_url: 'foo' } },
  ...failed
};
const disableTaskFailed = {
  id: '1',
  meta: { isTask: true, operation: 'disable' },
  context_data: { request_params: { vmware_vddk_package_url: 'foo' } },
  ...failed
};

describe('conversion hosts list item', () => {
  const getBaseProps = () => ({
    isTask: false,
    setHostToDeleteAction: jest.fn(),
    showConversionHostDeleteModalAction: jest.fn(),
    setConversionHostTaskToRetryAction: jest.fn(),
    showConversionHostRetryModalAction: jest.fn(),
    isPostingConversionHosts: false,
    saveTextFileAction: jest.fn()
  });

  it('renders correctly with an enabled conversion host', () => {
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={conversionHostEnabled} />);
    const actions = shallow(component.props().actions);
    expect(actions.find(ConversionHostRemoveButton).props().disabled).toBe(false);
    expect(
      actions
        .find(DropdownKebab)
        .find(MenuItem)
        .props().disabled
    ).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with a conversion host being disabled', () => {
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={conversionHostDisabling} />);
    const actions = shallow(component.props().actions);
    expect(actions.find(ConversionHostRemoveButton).props().disabled).toBe(true);
    expect(
      actions
        .find(DropdownKebab)
        .find(MenuItem)
        .props().disabled
    ).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with a manually enabled host (no enable task)', () => {
    const host = {
      name: 'Conversion Host 1',
      id: '1',
      meta: { tasksByOperation: {} }
    };
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={host} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with an in-progress enable task', () => {
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={enableTaskInProgress} isTask />);
    const actions = shallow(component.props().actions);
    expect(
      actions
        .find(DropdownKebab)
        .find(MenuItem)
        .props().disabled
    ).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with a failed enable task', () => {
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={enableTaskFailed} isTask />);
    const actions = shallow(component.props().actions);
    expect(actions.find(ConversionHostRetryButton)).toHaveLength(1);
    expect(
      actions
        .find(DropdownKebab)
        .find(MenuItem)
        .props().disabled
    ).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with a failed disable task', () => {
    const component = shallow(<ConversionHostsListItem {...getBaseProps()} listItem={disableTaskFailed} isTask />);
    const actions = shallow(component.props().actions);
    expect(actions.find(ConversionHostRetryButton)).toHaveLength(1);
    expect(
      actions
        .find(DropdownKebab)
        .find(MenuItem)
        .props().disabled
    ).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it('calls the correct action when downloading a log', () => {
    const baseProps = getBaseProps();
    const component = shallow(<ConversionHostsListItem {...baseProps} listItem={conversionHostEnabled} />);
    const actions = shallow(component.props().actions);
    actions
      .find(DropdownKebab)
      .find(MenuItem)
      .simulate('click');
    expect(baseProps.saveTextFileAction).toHaveBeenCalledTimes(1);
    const fileObject = baseProps.saveTextFileAction.mock.calls[0][0];
    expect(fileObject.fileName).toBeTruthy();
    expect(fileObject.fileBody).toBeTruthy();
  });

  it('does not call the save file action when trying to download a bogus log task', () => {
    const baseProps = getBaseProps();
    const task = { ...disableTaskFailed, meta: { isTask: true, operation: 'unknown' } };
    const component = shallow(<ConversionHostsListItem {...baseProps} listItem={task} isTask />);
    const actions = shallow(component.props().actions);
    actions
      .find(DropdownKebab)
      .find(MenuItem)
      .simulate('click');
    expect(baseProps.saveTextFileAction).toHaveBeenCalledTimes(0);
  });
});
