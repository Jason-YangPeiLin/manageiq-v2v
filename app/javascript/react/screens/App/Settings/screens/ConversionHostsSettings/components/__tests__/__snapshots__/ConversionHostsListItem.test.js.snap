// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion hosts list item renders correctly with a conversion host being disabled 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <ConversionHostRemoveButton
        disabled={true}
        host={
          Object {
            "id": "1",
            "meta": Object {
              "tasksByOperation": Object {
                "disable": Array [
                  Object {
                    "id": "3",
                    "message": "Example progress message",
                    "meta": Object {
                      "isTask": true,
                      "operation": "disable",
                    },
                    "state": "Active",
                    "status": "Ok",
                    "updated_on": "2019-05-13T18:07:11.849Z",
                  },
                  Object {
                    "context_data": Object {
                      "request_params": Object {
                        "mock": "params",
                      },
                    },
                    "id": "2",
                    "message": "Example error message",
                    "meta": Object {
                      "isTask": true,
                      "operation": "disable",
                    },
                    "state": "Finished",
                    "status": "Error",
                    "updated_on": "2019-05-13T18:06:11.849Z",
                  },
                ],
                "enable": Array [
                  Object {
                    "id": "1",
                    "message": "Example success message",
                    "meta": Object {
                      "isTask": true,
                      "operation": "enable",
                    },
                    "state": "Finished",
                    "status": "Ok",
                    "updated_on": "2019-05-13T18:05:11.849Z",
                  },
                ],
              },
            },
            "name": "Conversion Host 1",
            "vddk_transport_supported": true,
          }
        }
        setHostToDeleteAction={[MockFunction]}
        showConversionHostDeleteModalAction={[MockFunction]}
      />
      <StopPropagationOnClick>
        <DropdownKebab
          className=""
          componentClass={[Function]}
          id="task-kebab-3"
          pullRight={true}
          toggleStyle="link"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={true}
            divider={false}
            header={false}
            onClick={[Function]}
          >
            Download Log
          </MenuItem>
        </DropdownKebab>
      </StopPropagationOnClick>
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <Spinner
          className=""
          inline={true}
          inverse={false}
          loading={true}
          size="sm"
        />
        Removing...
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              <UNDEFINED>
                <strong>
                  Transformation method:
                </strong>
                 
                VDDK
                <br />
              </UNDEFINED>
              <UNDEFINED>
                <strong>
                  State:
                </strong>
                 
                Active
                <br />
                <strong>
                  Message:
                </strong>
                 
                Example progress message
              </UNDEFINED>
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading="Conversion Host 1"
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;

exports[`conversion hosts list item renders correctly with a failed disable task 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <UNDEFINED>
        <ConversionHostRetryButton
          disabled={false}
          setConversionHostTaskToRetryAction={[MockFunction]}
          showConversionHostRetryModalAction={[MockFunction]}
          task={
            Object {
              "context_data": Object {
                "request_params": Object {
                  "vmware_vddk_package_url": "foo",
                },
              },
              "id": "1",
              "message": "Example error message",
              "meta": Object {
                "isTask": true,
                "operation": "disable",
              },
              "state": "Finished",
              "status": "Error",
            }
          }
        />
      </UNDEFINED>
      <StopPropagationOnClick>
        <DropdownKebab
          className=""
          componentClass={[Function]}
          id="task-kebab-1"
          pullRight={true}
          toggleStyle="link"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            header={false}
            onClick={[Function]}
          >
            Download Log
          </MenuItem>
        </DropdownKebab>
      </StopPropagationOnClick>
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <ListViewIcon
          className=""
          name="error-circle-o"
          size="sm"
          type="pf"
        />
        Removal Failed
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              <UNDEFINED>
                <strong>
                  Transformation method:
                </strong>
                 
                VDDK
                <br />
              </UNDEFINED>
              <UNDEFINED>
                <strong>
                  State:
                </strong>
                 
                Finished
                <br />
                <strong>
                  Message:
                </strong>
                 
                Example error message
              </UNDEFINED>
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;

exports[`conversion hosts list item renders correctly with a failed enable task 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <UNDEFINED>
        <ConversionHostRetryButton
          disabled={false}
          setConversionHostTaskToRetryAction={[MockFunction]}
          showConversionHostRetryModalAction={[MockFunction]}
          task={
            Object {
              "context_data": Object {
                "request_params": Object {
                  "vmware_vddk_package_url": "foo",
                },
              },
              "id": "1",
              "message": "Example error message",
              "meta": Object {
                "isTask": true,
                "operation": "enable",
              },
              "state": "Finished",
              "status": "Error",
            }
          }
        />
      </UNDEFINED>
      <StopPropagationOnClick>
        <DropdownKebab
          className=""
          componentClass={[Function]}
          id="task-kebab-1"
          pullRight={true}
          toggleStyle="link"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            header={false}
            onClick={[Function]}
          >
            Download Log
          </MenuItem>
        </DropdownKebab>
      </StopPropagationOnClick>
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <ListViewIcon
          className=""
          name="error-circle-o"
          size="sm"
          type="pf"
        />
        Configuration Failed
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              <UNDEFINED>
                <strong>
                  Transformation method:
                </strong>
                 
                VDDK
                <br />
              </UNDEFINED>
              <UNDEFINED>
                <strong>
                  State:
                </strong>
                 
                Finished
                <br />
                <strong>
                  Message:
                </strong>
                 
                Example error message
              </UNDEFINED>
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;

exports[`conversion hosts list item renders correctly with a manually enabled host (no enable task) 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <ConversionHostRemoveButton
        disabled={undefined}
        host={
          Object {
            "id": "1",
            "meta": Object {
              "tasksByOperation": Object {},
            },
            "name": "Conversion Host 1",
          }
        }
        setHostToDeleteAction={[MockFunction]}
        showConversionHostDeleteModalAction={[MockFunction]}
      />
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <ListViewIcon
          className=""
          name="ok"
          size="sm"
          type="pf"
        />
        Configured
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              No configuration task information available
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading="Conversion Host 1"
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;

exports[`conversion hosts list item renders correctly with an enabled conversion host 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <ConversionHostRemoveButton
        disabled={false}
        host={
          Object {
            "id": "1",
            "meta": Object {
              "tasksByOperation": Object {
                "enable": Array [
                  Object {
                    "id": "1",
                    "message": "Example error message",
                    "meta": Object {
                      "isTask": true,
                      "operation": "enable",
                    },
                    "state": "Finished",
                    "status": "Error",
                    "updated_on": "2019-05-13T18:06:11.849Z",
                  },
                  Object {
                    "context_data": Object {
                      "conversion_host_check": "MOCK PLAYBOOK LOG OUTPUT",
                      "conversion_host_enable": "MOCK PLAYBOOK LOG OUTPUT",
                    },
                    "id": "2",
                    "message": "Example success message",
                    "meta": Object {
                      "isTask": true,
                      "operation": "enable",
                      "resourceName": "Conversion Host 1",
                    },
                    "state": "Finished",
                    "status": "Ok",
                    "updated_on": "2019-05-13T18:07:11.849Z",
                  },
                ],
              },
            },
            "name": "Conversion Host 1",
            "vddk_transport_supported": true,
          }
        }
        setHostToDeleteAction={[MockFunction]}
        showConversionHostDeleteModalAction={[MockFunction]}
      />
      <StopPropagationOnClick>
        <DropdownKebab
          className=""
          componentClass={[Function]}
          id="task-kebab-2"
          pullRight={true}
          toggleStyle="link"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            header={false}
            onClick={[Function]}
          >
            Download Log
          </MenuItem>
        </DropdownKebab>
      </StopPropagationOnClick>
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <ListViewIcon
          className=""
          name="ok"
          size="sm"
          type="pf"
        />
        Configured
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              <UNDEFINED>
                <strong>
                  Transformation method:
                </strong>
                 
                VDDK
                <br />
              </UNDEFINED>
              <UNDEFINED>
                <strong>
                  State:
                </strong>
                 
                Finished
                <br />
                <strong>
                  Message:
                </strong>
                 
                Example success message
              </UNDEFINED>
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading="Conversion Host 1"
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;

exports[`conversion hosts list item renders correctly with an in-progress enable task 1`] = `
<ListViewItem
  actions={
    <div
      className="conversion-hosts-list-actions"
    >
      <UNDEFINED>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          disabled={true}
        >
          Remove
        </Button>
      </UNDEFINED>
      <StopPropagationOnClick>
        <DropdownKebab
          className=""
          componentClass={[Function]}
          id="task-kebab-1"
          pullRight={true}
          toggleStyle="link"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={true}
            divider={false}
            header={false}
            onClick={[Function]}
          >
            Download Log
          </MenuItem>
        </DropdownKebab>
      </StopPropagationOnClick>
    </div>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <Spinner
          className=""
          inline={true}
          inverse={false}
          loading={true}
          size="sm"
        />
        Configuring...
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Popover
              bsClass="popover"
              id="task-info-popover"
              placement="right"
              style={
                Object {
                  "width": 400,
                }
              }
            >
              <UNDEFINED>
                <strong>
                  Transformation method:
                </strong>
                 
                VDDK
                <br />
              </UNDEFINED>
              <UNDEFINED>
                <strong>
                  State:
                </strong>
                 
                Active
                <br />
                <strong>
                  Message:
                </strong>
                 
                Example progress message
              </UNDEFINED>
            </Popover>
          }
          placement="top"
          rootClose={true}
          trigger="click"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="link"
            disabled={false}
          >
            <Icon
              name="info"
              type="pf"
            />
          </Button>
        </OverlayTrigger>
      </ListViewInfoItem>,
    ]
  }
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;
